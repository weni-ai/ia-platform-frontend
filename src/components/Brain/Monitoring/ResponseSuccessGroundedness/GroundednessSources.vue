<template>
  <section
    v-for="fragment in fragments"
    :key="fragment.sentence"
    class="groundedness-sources"
  >
    <section class="groundedness-sources__source">
      <span
        :class="[
          'source__status',
          `source__status--${getReliabilityLevel(fragment)}`,
        ]"
      />
      <UnnnicIntelligenceText
        v-if="fragment.sources[0].filename"
        class="source__filename"
        color="neutral-dark"
        family="secondary"
        size="body-gt"
        tag="p"
      >
        {{ fragment.sources[0].filename }}
      </UnnnicIntelligenceText>
      <p
        v-else
        class="source__filename--not-found"
      >
        {{ $t('router.monitoring.inspect_response.not_found') }}
      </p>
      <UnnnicIntelligenceText
        class="source__confidence"
        color="neutral-cloudy"
        family="secondary"
        size="body-gt"
        tag="p"
      >
        {{
          $t('router.monitoring.inspect_response.confidence_percentage', {
            percentage: fragment.score,
          })
        }}
      </UnnnicIntelligenceText>
    </section>
  </section>
</template>

<script setup>
defineProps({
  fragments: {
    type: Array,
    required: true,
  },
  getReliabilityLevel: {
    type: Function,
    required: true,
  },
});
</script>

<style scoped lang="scss">
.groundedness-sources {
  display: grid;
  gap: $unnnic-spacing-xs;

  .groundedness-sources__source {
    display: grid;
    grid-template-columns: auto auto 1fr;
    gap: $unnnic-spacing-xs;
    align-items: center;

    .source__status {
      width: 10px;
      height: 10px;
      border-radius: 10px;
      background-color: $unnnic-color-neutral-cleanest;

      &--slightly-reliable {
        background-color: $unnnic-color-weni-300;
      }
      &--moderately-reliable {
        background-color: $unnnic-color-weni-600;
      }
      &--highly-reliable {
        background-color: $unnnic-color-weni-800;
      }
    }

    .source__filename {
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;

      &--not-found {
        color: $unnnic-color-neutral-dark;
        font-family: $unnnic-font-family-secondary;
        font-size: $unnnic-font-size-body-gt;
        line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
        font-style: italic;
      }
    }

    .source__confidence {
      white-space: nowrap;
    }
  }
}
</style>
