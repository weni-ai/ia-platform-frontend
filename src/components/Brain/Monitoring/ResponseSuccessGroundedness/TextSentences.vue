<template>
  <section
    data-testid="sentences"
    class="text-sentences"
    @mouseleave="clearHoveredSentence"
  >
    <UnnnicToolTip
      v-for="(fragment, index) in fragments"
      :key="fragment.sentence"
      data-testid="sentence-text"
      :text="getTooltipText(fragment)"
      :class="getTextClasses(index, fragment)"
      side="top"
      enabled
      :forceOpen="hoveredSentence === index"
      @mouseover="setHoveredSentence(index)"
    >
      {{ fragment.sentence }}
    </UnnnicToolTip>
  </section>
</template>

<script setup>
import { ref } from 'vue';

import i18n from '@/utils/plugins/i18n';

const props = defineProps({
  fragments: {
    type: Array,
    required: true,
  },
  getReliabilityLevel: {
    type: Function,
    required: true,
  },
});

const hoveredSentence = ref(null);

function setHoveredSentence(index) {
  hoveredSentence.value = index;
}

function clearHoveredSentence() {
  hoveredSentence.value = null;
}

function truncateString(string, maxLength = 25) {
  if (!string) return;

  return string.length > maxLength
    ? `${string.slice(0, maxLength)}...`
    : string;
}

function getTooltipText(fragment) {
  const truncatedFileName = truncateString(fragment.sources?.[0].filename);
  const confidence = i18n.global.t(
    'router.monitoring.inspect_response.confidence_percentage',
    {
      percentage: fragment.score,
    },
  );
  return `${truncatedFileName}\n(${confidence})`;
}

function getTextClasses(index, fragment) {
  return [
    'text-sentences__text',
    `text-sentences__text--${props.getReliabilityLevel(fragment)}`,
    { 'text-sentences__text--focused': hoveredSentence.value === index },
  ];
}
</script>

<style scoped lang="scss">
.text-sentences {
  .text-sentences__text {
    margin-right: $unnnic-spacing-nano;

    display: inline;

    line-height: $unnnic-font-size-body-lg + $unnnic-line-height-md;
    color: $unnnic-color-neutral-darkest;
    font-size: $unnnic-font-size-body-lg;
    text-decoration: underline;
    text-decoration-color: $unnnic-color-neutral-cleanest;
    text-decoration-thickness: 3px;
    text-underline-offset: 4px;

    cursor: default;

    &--focused {
      text-decoration-thickness: 5px;
    }

    &--slightly-reliable {
      text-decoration-color: $unnnic-color-weni-300;
    }
    &--moderately-reliable {
      text-decoration-color: $unnnic-color-weni-600;
    }
    &--highly-reliable {
      text-decoration-color: $unnnic-color-weni-800;
    }
  }
}
</style>
