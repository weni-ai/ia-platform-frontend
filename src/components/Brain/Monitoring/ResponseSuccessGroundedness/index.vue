<template>
  <section
    class="response-success-groundedness"
    data-testid="response-success-groundedness"
  >
    <TextSentences
      :fragments="groundednessFragments"
      :getReliabilityLevel="getReliabilityLevel"
    />
    <GroundednessSources
      :fragments="groundednessFragments"
      :getReliabilityLevel="getReliabilityLevel"
    />
  </section>
</template>

<script setup>
import { computed } from 'vue';

import TextSentences from './TextSentences.vue';
import GroundednessSources from './GroundednessSources.vue';

const props = defineProps({
  inspectionData: {
    type: Object,
    required: true,
  },
});

const groundednessFragments = computed(() => {
  const data = props.inspectionData.groundedness.map((fragment) => ({
    ...fragment,
  }));

  if (data.length > 0) {
    const firstFragment = data[0];
    const lastFragment = data.at(-1);

    firstFragment.sentence = `“${firstFragment.sentence}`;
    lastFragment.sentence = `${lastFragment.sentence}”`;
  }

  return data;
});

function getReliabilityLevel(item) {
  const { score } = item;

  if (score === 0) return 'unreliable';
  if (score <= 33) return 'slightly-reliable';
  if (score <= 66) return 'moderately-reliable';
  if (score <= 100) return 'highly-reliable';
  return 'unreliable';
}
</script>

<style scoped lang="scss">
.response-success-groundedness {
  display: grid;
  gap: $unnnic-spacing-sm;
}
</style>
