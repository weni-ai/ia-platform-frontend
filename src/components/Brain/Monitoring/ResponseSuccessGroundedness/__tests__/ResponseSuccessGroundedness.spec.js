import { describe, it, expect, beforeEach, vi } from 'vitest';
import { shallowMount } from '@vue/test-utils';
import ResponseSuccessGroundedness from '../index.vue';
import TextSentences from '../TextSentences.vue';
import GroundednessSources from '../GroundednessSources.vue';

describe('ResponseSuccessGroundedness', () => {
  let wrapper;

  const inspectionDataMock = {
    groundedness: [
      { sentence: 'First sentence', score: 85 },
      { sentence: 'Second sentence', score: 50 },
      { sentence: 'Third sentence', score: 15 },
    ],
  };

  const createWrapper = (props = {}) => {
    wrapper = shallowMount(ResponseSuccessGroundedness, {
      props: {
        inspectionData: inspectionDataMock,
        ...props,
      },
    });
  };

  beforeEach(() => {
    createWrapper();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders the component successfully', () => {
    const responseSuccessGroundedness = wrapper.find(
      '[data-testid="response-success-groundedness"]',
    );
    expect(responseSuccessGroundedness.exists()).toBe(true);
  });

  it('passes correctly props to TextSentences', () => {
    const textSentences = wrapper.findComponent(TextSentences);
    expect(textSentences.exists()).toBe(true);

    expect(textSentences.props('fragments')).toEqual(
      wrapper.vm.groundednessFragments,
    );
  });

  it('passes correctly props to GroundednessSources', () => {
    const groundednessSources = wrapper.findComponent(GroundednessSources);
    expect(groundednessSources.exists()).toBe(true);

    expect(groundednessSources.props('fragments')).toEqual(
      wrapper.vm.groundednessFragments,
    );
  });

  it('computes groundednessFragments correctly', () => {
    const groundednessFragments = wrapper.vm.groundednessFragments;

    const expectedFragments = inspectionDataMock.groundedness.map(
      (fragment, index) => ({
        ...fragment,
        sentence:
          index === 0
            ? `“${fragment.sentence}`
            : index === inspectionDataMock.groundedness.length - 1
              ? `${fragment.sentence}”`
              : fragment.sentence,
      }),
    );

    expect(groundednessFragments).toEqual(expectedFragments);
  });

  it('provides the correct getReliabilityLevel function to child components', () => {
    const groundednessSources = wrapper.findComponent(GroundednessSources);
    expect(groundednessSources.props('getReliabilityLevel')).toBe(
      wrapper.vm.getReliabilityLevel,
    );

    const reliabilityLevels = [
      { score: 0, expected: 'unreliable' },
      { score: 20, expected: 'slightly-reliable' },
      { score: 50, expected: 'moderately-reliable' },
      { score: 85, expected: 'highly-reliable' },
      { score: 110, expected: 'unreliable' },
    ];

    reliabilityLevels.forEach(({ score, expected }) => {
      const result = wrapper.vm.getReliabilityLevel({ score });
      expect(result).toBe(expected);
    });
  });
});
