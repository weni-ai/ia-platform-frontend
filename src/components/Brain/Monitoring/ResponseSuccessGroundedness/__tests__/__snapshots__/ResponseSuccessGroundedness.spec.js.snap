// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResponseSuccessGroundedness > matches the snapshot 1`] = `
"<section data-v-cd526e99="" class="response-success-groundedness" data-testid="response-success-groundedness">
  <text-sentences-stub data-v-cd526e99="" fragments="[object Object],[object Object],[object Object]" getreliabilitylevel="[Function]"></text-sentences-stub>
  <groundedness-sources-stub data-v-cd526e99="" fragments="[object Object],[object Object],[object Object]" getreliabilitylevel="[Function]"></groundedness-sources-stub>
</section>"
`;
